apps/meteor/server/services/calendar/statusEvents.ts

```diff
@@ -0,0 +1,202 @@
+				// Event ends during our event
+				{ endTime: { $gt: startTime, $lte: endTime } },
+				// Event completely contains our event
+				{ startTime: { $lte: startTime }, endTime: { $gte: endTime } },
+			],
+		}).toArray();
+
+		if (overlappingEvents.length === 0) {
+			return { shouldProceed: true };
+		}
+
+		const allEvents = [...overlappingEvents, { endTime, startTime }];
+
+		const latestEndingEvent = allEvents.reduce<{ endTime: Date | null; startTime: Date | null }>(
+			(latest, event) => {
+				if (!event.endTime) return latest;
+				if (!latest.endTime || event.endTime > latest.endTime) {
+					return { endTime: event.endTime, startTime: event.startTime };
+				}
+				return latest;
+			},
+			{ endTime: null, startTime: null },
+		);
+
+		// If this event doesn't have the latest end time, don't schedule removal
+		// because another event will handle it
+		if (latestEndingEvent.endTime && latestEndingEvent.endTime.getTime() !== endTime.getTime()) {
+			const scheduledTime = startTime;
+			const cronJobId = this.generateCronJobId(eventId, uid, 'status');
+
+			if (await cronJobs.has(cronJobId)) {
+				await cronJobs.remove(cronJobId);
+			}
+
+			await cronJobs.addAtTimestamp(cronJobId, scheduledTime, async () =>
+				this.applyStatusChange(eventId, uid, startTime, endTime, status, false),
+			);
+			return { shouldProceed: false };
+		}
+
+		// For any existing events that end before this one, remove their status removal jobs
+		for await (const event of overlappingEvents) {
+			if (event.endTime && event.endTime < endTime) {
+				const eventCronJobId = this.generateCronJobId(event._id, uid, 'status');
+				if (await cronJobs.has(eventCronJobId)) {
+					await cronJobs.remove(eventCronJobId);
+				}
+			}
+		}
+
+		return { shouldProceed: true };
+	}
+
+	public async setupAppointmentStatusChange(
+		eventId: ICalendarEvent['_id'],
+		uid: IUser['_id'],
+		startTime: Date,
+		endTime?: Date,
+		status?: UserStatus,
+		shouldScheduleRemoval?: boolean,
+	): Promise<void> {
+		const hasBusyStatusSetting = settings.get<boolean>('Calendar_BusyStatus_Enabled');
+		if (!endTime || !hasBusyStatusSetting) {
+			return;
+		}
+
+		if (shouldScheduleRemoval) {
+			const { shouldProceed } = await this.handleOverlappingEvents(eventId, uid, startTime, endTime, status);
+			if (!shouldProceed) {
+				return;
+			}
+		}
+
+		const scheduledTime = shouldScheduleRemoval ? startTime : endTime;
+		const cronJobId = this.generateCronJobId(eventId, uid, 'status');
+
+		if (await cronJobs.has(cronJobId)) {
+			await cronJobs.remove(cronJobId);
+		}
+
+		await cronJobs.addAtTimestamp(cronJobId, scheduledTime, async () =>
+			this.applyStatusChange(eventId, uid, startTime, endTime, status, shouldScheduleRemoval),
+		);
+	}
+
+	public async applyStatusChange(
+		eventId: ICalendarEvent['_id'],
+		uid: IUser['_id'],
+		startTime: Date,
+		endTime?: Date,
+		status?: UserStatus,
+		shouldScheduleRemoval?: boolean,
```

Comment:
Can we change the method parameters to an object? Mainly because we have a boolean param, it's hard to know what boolean params are for.